lib/guess_html_encoding.rb

```diff
@@ -19,11 +19,20 @@ def self.guess(html, headers = nil)
     end
 
     if out.nil? || out.empty? || !encoding_loaded?(out)
-      if html =~ /<meta[^>]*HTTP-EQUIV=["']?Content-Type["']?[^>]*content=["']([^'"]*)["']/i && $1 =~ /charset=([\w\d-]+);?/i
-        out = $1
-      elsif html =~ /<meta\s+charset=["']([\w\d-]+)?/i
-        out = $1
+
+      meta_tags_regex = /
+      (?:
+        <meta[^>]*HTTP-EQUIV=["']?Content-Type["']?[^>]*content=["'][^'"]*charset\s*=\s*([^"'\;\s]+)  # http-equiv
+        |                                                                                              # or
```

Comment:
@cantino good point. I've changed this again, to allow for any other attributes before the `http-equiv=` (with a test), and also made the charset match any substring of characters leading up to a `;` or space or quote mark.

The full algorithm is pretty comprehensively specified in the HTML5 spec: http://www.w3.org/html/wg/drafts/html/master/infrastructure.html#algorithm-for-extracting-a-character-encoding-from-a-meta-element – and also allows for the `content=` attribute to appear before the `http-equiv=`, which we don't do yet, and for the `content=` to either use matching `"` or `'` marks, or no marks at all.

So I think all of the following are technically valid:

    <meta http-equiv='content-type' content='text/html; charset=LATIN1;'>
    <meta http-equiv="content-type" content="text/html; charset=LATIN1">
    <meta http-equiv="content-type" content=text/html; charset=LATIN1;>
    <meta content='text/html; charset=LATIN1;' http-equiv='content-type'>

but I'm not 100% sure. To get closer to the spec, I think we'd have to either implement (or use existing cod for) the attribute-scanning algorithm mentioned here: http://www.w3.org/html/wg/drafts/html/master/syntax.html#concept-get-attributes-when-sniffing – but I'm not sure how far down that rabbit hole you want to take this gem? :)



but I'm not 100% sure.